File: Sources/PalmierPro/Utilities/TimeFormatting.swift
func formatTimecode(frame: Int, fps: Int) -> String {
    guard fps > 0 else { return "00:00:00:00" }
    let totalSeconds = frame / fps
    let ff = frame % fps
    let ss = totalSeconds % 60
    let mm = (totalSeconds / 60) % 60
    let hh = totalSeconds / 3600
    return "\(twoDigit(hh)):\(twoDigit(mm)):\(twoDigit(ss)):\(twoDigit(ff))"
}

private func twoDigit(_ value: Int) -> String {
    guard value >= 0 && value < 10 else { return "\(value)" }
    return "0\(value)"
}

func frameToSeconds(frame: Int, fps: Int) -> Double {
    guard fps > 0 else { return 0 }
    return Double(frame) / Double(fps)
}

func secondsToFrame(seconds: Double, fps: Int) -> Int {
    Int(seconds * Double(fps))
}

File: platforms/windows/src/PalmierPro.Core/TimeFormatting.cs
namespace PalmierPro.Core;

/// Ported from Utilities/TimeFormatting.swift's `formatTimecode(frame:fps:)` — always the full
/// HH:MM:SS:FF form, unlike TimelineCanvasControl's compact ruler-label variant (which drops a
/// leading all-zero hours group).
public static class TimeFormatting
{
    public static string FormatTimecode(int frame, int fps)
    {
        if (fps <= 0)
        {
            return "00:00:00:00";
        }
        var absFrame = Math.Abs(frame);
        var totalSeconds = absFrame / fps;
        var ff = absFrame % fps;
        var ss = totalSeconds % 60;
        var mm = totalSeconds / 60 % 60;
        var hh = totalSeconds / 3600;
        var sign = frame < 0 ? "-" : "";
        return $"{sign}{hh:D2}:{mm:D2}:{ss:D2}:{ff:D2}";
    }
}
